File: scripts/segmented_lidar_cluster_node.py
#!/usr/bin/env python3

from collections import defaultdict, deque
import math

import cv2
import message_filters
import numpy as np
import rclpy
from cv_bridge import CvBridge
from geometry_msgs.msg import Pose, PoseArray
from rclpy.duration import Duration
from rclpy.node import Node
from rclpy.qos import QoSProfile, ReliabilityPolicy
from sensor_msgs.msg import CameraInfo, CompressedImage, PointCloud2
from sensor_msgs_py import point_cloud2
from tf2_ros import Buffer, TransformException, TransformListener
from visualization_msgs.msg import Marker, MarkerArray


class SegmentedLidarClusterNode(Node):
    def __init__(self):
        super().__init__("segmented_lidar_cluster_node")
        self.bridge = CvBridge()
        self.camera_info = None
        self.tf_buffer = Buffer()
        self.tf_listener = TransformListener(self.tf_buffer, self)

        self.declare_parameter("mask_topic", "/sam2_segmentation_node/binary_mask/compressed")
        self.declare_parameter("pointcloud_topic", "/lidar/points")
        self.declare_parameter("camera_info_topic", "/camera/camera_head/color/camera_info")
        self.declare_parameter("filtered_cloud_topic", "/segmented_lidar/points")
        self.declare_parameter("centers_topic", "/segmented_lidar/cluster_centers")
        self.declare_parameter("markers_topic", "/segmented_lidar/cluster_markers")
        self.declare_parameter("camera_frame", "")
        self.declare_parameter("mask_threshold", 0)
        self.declare_parameter("cluster_tolerance", 0.30)
        self.declare_parameter("min_cluster_size", 8)
        self.declare_parameter("max_cluster_size", 30000)
        self.declare_parameter("sync_queue_size", 10)
        self.declare_parameter("sync_slop", 0.10)
        self.declare_parameter("tf_timeout_sec", 0.05)

        self.mask_topic = self.get_parameter("mask_topic").value
        self.pointcloud_topic = self.get_parameter("pointcloud_topic").value
        self.camera_info_topic = self.get_parameter("camera_info_topic").value
        self.camera_frame_param = self.get_parameter("camera_frame").value
        self.mask_threshold = int(self.get_parameter("mask_threshold").value)
        self.cluster_tolerance = float(self.get_parameter("cluster_tolerance").value)
        self.min_cluster_size = int(self.get_parameter("min_cluster_size").value)
        self.max_cluster_size = int(self.get_parameter("max_cluster_size").value)
        self.tf_timeout = Duration(seconds=float(self.get_parameter("tf_timeout_sec").value))

        qos = QoSProfile(depth=10, reliability=ReliabilityPolicy.BEST_EFFORT)
        reliable_qos = QoSProfile(depth=10, reliability=ReliabilityPolicy.RELIABLE)

        self.camera_info_sub = self.create_subscription(
            CameraInfo,
            self.camera_info_topic,
            self.camera_info_callback,
            reliable_qos,
        )

        self.cloud_pub = self.create_publisher(
            PointCloud2,
            self.get_parameter("filtered_cloud_topic").value,
            qos,
        )
        self.centers_pub = self.create_publisher(
            PoseArray,
            self.get_parameter("centers_topic").value,
            reliable_qos,
        )
        self.markers_pub = self.create_publisher(
            MarkerArray,
            self.get_parameter("markers_topic").value,
            reliable_qos,
        )

        self.mask_sub = message_filters.Subscriber(self, CompressedImage, self.mask_topic, qos_profile=qos)
        self.cloud_sub = message_filters.Subscriber(
            self,
            PointCloud2,
            self.pointcloud_topic,
            qos_profile=qos,
        )
        self.sync = message_filters.ApproximateTimeSynchronizer(
            [self.mask_sub, self.cloud_sub],
            queue_size=int(self.get_parameter("sync_queue_size").value),
            slop=float(self.get_parameter("sync_slop").value),
        )
        self.sync.registerCallback(self.synced_callback)

        self.get_logger().info(
            "Segmented LiDAR clustering node started. "
            f"mask={self.mask_topic}, cloud={self.pointcloud_topic}, camera_info={self.camera_info_topic}"
        )

    def camera_info_callback(self, msg):
        self.camera_info = msg

    def synced_callback(self, mask_msg, cloud_msg):
        if self.camera_info is None:
            self.get_logger().warn("Waiting for CameraInfo.", throttle_duration_sec=2.0)
            return

        camera_frame = self.camera_frame_param or self.camera_info.header.frame_id
        if not camera_frame:
            self.get_logger().warn("Camera frame is empty. Set camera_frame or publish CameraInfo frame_id.")
            return

        mask = self._mask_to_binary(mask_msg)
        if mask is None:
            return

        try:
            transform = self.tf_buffer.lookup_transform(
                camera_frame,
                cloud_msg.header.frame_id,
                cloud_msg.header.stamp,
                timeout=self.tf_timeout,
            )
        except TransformException as exc:
            self.get_logger().warn(
                f"Cannot transform cloud frame '{cloud_msg.header.frame_id}' to camera frame "
                f"'{camera_frame}': {exc}",
                throttle_duration_sec=2.0,
            )
            return

        lidar_points = self._read_xyz_points(cloud_msg)
        if lidar_points.size == 0:
            self._publish_empty(cloud_msg)
            return

        camera_points = self._transform_points(lidar_points, transform)
        selected_points = self._select_points_on_mask(lidar_points, camera_points, mask)
        clusters = self._cluster_points(selected_points)
        centers = self._cluster_centers(selected_points, clusters)

        self._publish_filtered_cloud(cloud_msg, selected_points)
        self._publish_centers(cloud_msg.header, centers)
        self._publish_markers(cloud_msg.header, centers, clusters, selected_points)

    def _mask_to_binary(self, mask_msg):
        try:
            mask = self.bridge.compressed_imgmsg_to_cv2(mask_msg, desired_encoding="mono8")
        except Exception as exc:
            self.get_logger().warn(f"Failed to convert mask image: {exc}", throttle_duration_sec=2.0)
            return None

        if mask.ndim == 3:
            mask = cv2.cvtColor(mask, cv2.COLOR_BGR2GRAY)
        return mask > self.mask_threshold

    def _read_xyz_points(self, cloud_msg):
        points = point_cloud2.read_points(
            cloud_msg,
            field_names=("x", "y", "z"),
            skip_nans=True,
        )
        if isinstance(points, np.ndarray):
            if points.dtype.names:
                array = np.column_stack((points["x"], points["y"], points["z"])).astype(np.float32)
            else:
                array = np.asarray(points, dtype=np.float32)
        else:
            array = np.asarray(list(points), dtype=np.float32)

        if array.size == 0:
            return np.empty((0, 3), dtype=np.float32)
        return array.reshape((-1, 3))

    def _transform_points(self, points, transform):
        rotation = self._quaternion_to_matrix(transform.transform.rotation)
        translation = np.array(
            [
                transform.transform.translation.x,
                transform.transform.translation.y,
                transform.transform.translation.z,
            ],
            dtype=np.float32,
        )
        return points @ rotation.T + translation

    def _select_points_on_mask(self, lidar_points, camera_points, mask):
        fx = self.camera_info.k[0]
        fy = self.camera_info.k[4]
        cx = self.camera_info.k[2]
        cy = self.camera_info.k[5]

        z = camera_points[:, 2]
        valid_depth = z > 0.0
        if not np.any(valid_depth):
            return np.empty((0, 3), dtype=np.float32)

        projected_u = np.zeros(camera_points.shape[0], dtype=np.int32)
        projected_v = np.zeros(camera_points.shape[0], dtype=np.int32)
        valid_indices = np.where(valid_depth)[0]

        projected_u[valid_indices] = np.round(
            fx * camera_points[valid_indices, 0] / z[valid_indices] + cx
        ).astype(np.int32)
        projected_v[valid_indices] = np.round(
            fy * camera_points[valid_indices, 1] / z[valid_indices] + cy
        ).astype(np.int32)

        height, width = mask.shape[:2]
        if self.camera_info.width > 0 and self.camera_info.height > 0:
            projected_u = np.round(projected_u * (width / self.camera_info.width)).astype(np.int32)
            projected_v = np.round(projected_v * (height / self.camera_info.height)).astype(np.int32)

        in_image = (
            valid_depth
            & (projected_u >= 0)
            & (projected_u < width)
            & (projected_v >= 0)
            & (projected_v < height)
        )
        if not np.any(in_image):
            return np.empty((0, 3), dtype=np.float32)

        image_indices = np.where(in_image)[0]
        on_mask = mask[projected_v[image_indices], projected_u[image_indices]]
        selected_indices = image_indices[on_mask]
        return lidar_points[selected_indices]

    def _cluster_points(self, points):
        if points.shape[0] < self.min_cluster_size:
            return []

        cell_size = self.cluster_tolerance
        cells = defaultdict(list)
        for index, point in enumerate(points):
            key = tuple(np.floor(point / cell_size).astype(np.int32).tolist())
            cells[key].append(index)

        visited = np.zeros(points.shape[0], dtype=bool)
        clusters = []

        for start_index in range(points.shape[0]):
            if visited[start_index]:
                continue

            cluster = []
            queue = deque([start_index])
            visited[start_index] = True

            while queue:
                current = queue.popleft()
                cluster.append(current)
                current_key = tuple(np.floor(points[current] / cell_size).astype(np.int32).tolist())

                for neighbor_key in self._neighbor_cell_keys(current_key):
                    for neighbor in cells.get(neighbor_key, []):
                        if visited[neighbor]:
                            continue
                        if np.linalg.norm(points[current] - points[neighbor]) <= self.cluster_tolerance:
                            visited[neighbor] = True
                            queue.append(neighbor)

            if self.min_cluster_size <= len(cluster) <= self.max_cluster_size:
                clusters.append(np.array(cluster, dtype=np.int32))

        return clusters

    def _neighbor_cell_keys(self, key):
        for dx in (-1, 0, 1):
            for dy in (-1, 0, 1):
                for dz in (-1, 0, 1):
                    yield key[0] + dx, key[1] + dy, key[2] + dz

    def _cluster_centers(self, points, clusters):
        centers = []
        for cluster in clusters:
            centers.append(np.mean(points[cluster], axis=0))
        return centers

    def _publish_filtered_cloud(self, source_msg, selected_points):
        cloud = point_cloud2.create_cloud_xyz32(source_msg.header, selected_points.tolist())
        self.cloud_pub.publish(cloud)

    def _publish_centers(self, header, centers):
        pose_array = PoseArray()
        pose_array.header = header

        for center in centers:
            pose = Pose()
            pose.position.x = float(center[0])
            pose.position.y = float(center[1])
            pose.position.z = float(center[2])
            pose.orientation.w = 1.0
            pose_array.poses.append(pose)

        self.centers_pub.publish(pose_array)

    def _publish_markers(self, header, centers, clusters, points):
        marker_array = MarkerArray()

        delete_marker = Marker()
        delete_marker.header = header
        delete_marker.action = Marker.DELETEALL
        marker_array.markers.append(delete_marker)

        for marker_id, (center, cluster) in enumerate(zip(centers, clusters)):
            cluster_points = points[cluster]
            min_xyz = np.min(cluster_points, axis=0)
            max_xyz = np.max(cluster_points, axis=0)
            extent = np.maximum(max_xyz - min_xyz, 0.05)

            marker = Marker()
            marker.header = header
            marker.ns = "segmented_lidar_clusters"
            marker.id = marker_id
            marker.type = Marker.CUBE
            marker.action = Marker.ADD
            marker.pose.position.x = float(center[0])
            marker.pose.position.y = float(center[1])
            marker.pose.position.z = float(center[2])
            marker.pose.orientation.w = 1.0
            marker.scale.x = float(extent[0])
            marker.scale.y = float(extent[1])
            marker.scale.z = float(extent[2])
            marker.color.r = 0.1
            marker.color.g = 0.8
            marker.color.b = 1.0
            marker.color.a = 0.35
            marker.lifetime.sec = 0
            marker_array.markers.append(marker)

            text_marker = Marker()
            text_marker.header = header
            text_marker.ns = "segmented_lidar_cluster_labels"
            text_marker.id = marker_id
            text_marker.type = Marker.TEXT_VIEW_FACING
            text_marker.action = Marker.ADD
            text_marker.pose.position.x = float(center[0])
            text_marker.pose.position.y = float(center[1])
            text_marker.pose.position.z = float(center[2] + extent[2] * 0.5 + 0.15)
            text_marker.pose.orientation.w = 1.0
            text_marker.scale.z = 0.18
            text_marker.color.r = 1.0
            text_marker.color.g = 1.0
            text_marker.color.b = 1.0
            text_marker.color.a = 1.0
            text_marker.text = (
                f"{marker_id}: ({center[0]:.2f}, {center[1]:.2f}, {center[2]:.2f}) "
                f"n={len(cluster)}"
            )
            marker_array.markers.append(text_marker)

        self.markers_pub.publish(marker_array)

    def _publish_empty(self, source_msg):
        self._publish_filtered_cloud(source_msg, np.empty((0, 3), dtype=np.float32))
        self._publish_centers(source_msg.header, [])
        self._publish_markers(source_msg.header, [], [], np.empty((0, 3), dtype=np.float32))

    def _quaternion_to_matrix(self, quat):
        x = quat.x
        y = quat.y
        z = quat.z
        w = quat.w
        norm = math.sqrt(x * x + y * y + z * z + w * w)
        if norm == 0.0:
            return np.eye(3, dtype=np.float32)

        x /= norm
        y /= norm
        z /= norm
        w /= norm

        return np.array(
            [
                [1.0 - 2.0 * (y * y + z * z), 2.0 * (x * y - z * w), 2.0 * (x * z + y * w)],
                [2.0 * (x * y + z * w), 1.0 - 2.0 * (x * x + z * z), 2.0 * (y * z - x * w)],
                [2.0 * (x * z - y * w), 2.0 * (y * z + x * w), 1.0 - 2.0 * (x * x + y * y)],
            ],
            dtype=np.float32,
        )


def main(args=None):
    rclpy.init(args=args)
    node = SegmentedLidarClusterNode()
    try:
        rclpy.spin(node)
    finally:
        node.destroy_node()
        rclpy.shutdown()


if __name__ == "__main__":
    main()

File: src/segmented_lidar_cluster_node.cpp
#include <algorithm>
#include <cmath>
#include <cstdint>
#include <cstdio>
#include <functional>
#include <memory>
#include <string>
#include <vector>

#include <Eigen/Geometry>
#include <cv_bridge/cv_bridge.h>
#include <geometry_msgs/msg/pose.hpp>
#include <geometry_msgs/msg/pose_array.hpp>
#include <message_filters/subscriber.h>
#include <message_filters/sync_policies/approximate_time.h>
#include <message_filters/synchronizer.h>
#include <opencv2/imgproc.hpp>
#include <pcl/common/centroid.h>
#include <pcl/common/common.h>
#include <pcl/common/point_tests.h>
#include <pcl/common/transforms.h>
#include <pcl/kdtree/kdtree.h>
#include <pcl/point_cloud.h>
#include <pcl/point_types.h>
#include <pcl/segmentation/extract_clusters.h>
#include <pcl_conversions/pcl_conversions.h>
#include <rclcpp/rclcpp.hpp>
#include <sensor_msgs/msg/camera_info.hpp>
#include <sensor_msgs/msg/compressed_image.hpp>
#include <sensor_msgs/msg/point_cloud2.hpp>
#include <std_msgs/msg/header.hpp>
#include <tf2/exceptions.h>
#include <tf2_ros/buffer.h>
#include <tf2_ros/transform_listener.h>
#include <visualization_msgs/msg/marker.hpp>
#include <visualization_msgs/msg/marker_array.hpp>

class SegmentedLidarClusterNode : public rclcpp::Node {
public:
  SegmentedLidarClusterNode()
      : Node("segmented_lidar_cluster_node"), tf_buffer_(this->get_clock()),
        tf_listener_(tf_buffer_) {
    mask_topic_ = this->declare_parameter<std::string>("mask_topic",
                                                       "/sam2_segmentation_node/binary_mask/compressed");
    pointcloud_topic_ = this->declare_parameter<std::string>("pointcloud_topic",
                                                             "/livox/lidar");
    camera_info_topic_ = this->declare_parameter<std::string>(
        "camera_info_topic", "/camera/camera_head/color/camera_info");
    filtered_cloud_topic_ = this->declare_parameter<std::string>(
        "filtered_cloud_topic", "/segmented_lidar/points");
    centers_topic_ = this->declare_parameter<std::string>(
        "centers_topic", "/segmented_lidar/cluster_centers");
    markers_topic_ = this->declare_parameter<std::string>(
        "markers_topic", "/segmented_lidar/cluster_markers");
    camera_frame_param_ =
        this->declare_parameter<std::string>("camera_frame", "");
    mask_threshold_ = this->declare_parameter<int>("mask_threshold", 0);
    cluster_tolerance_ =
        this->declare_parameter<double>("cluster_tolerance", 0.30);
    min_cluster_size_ = this->declare_parameter<int>("min_cluster_size", 8);
    max_cluster_size_ =
        this->declare_parameter<int>("max_cluster_size", 30000);
    const int sync_queue_size =
        this->declare_parameter<int>("sync_queue_size", 10);
    const double sync_slop = this->declare_parameter<double>("sync_slop", 0.10);
    tf_timeout_sec_ = this->declare_parameter<double>("tf_timeout_sec", 0.05);

    auto best_effort_qos =
        rclcpp::QoS(rclcpp::KeepLast(10)).best_effort();
    auto reliable_qos = rclcpp::QoS(rclcpp::KeepLast(10)).reliable();

    camera_info_sub_ = this->create_subscription<sensor_msgs::msg::CameraInfo>(
        camera_info_topic_, reliable_qos,
        std::bind(&SegmentedLidarClusterNode::cameraInfoCallback, this,
                  std::placeholders::_1));

    filtered_cloud_pub_ =
        this->create_publisher<sensor_msgs::msg::PointCloud2>(
            filtered_cloud_topic_, best_effort_qos);
    centers_pub_ = this->create_publisher<geometry_msgs::msg::PoseArray>(
        centers_topic_, reliable_qos);
    markers_pub_ = this->create_publisher<visualization_msgs::msg::MarkerArray>(
        markers_topic_, reliable_qos);

    mask_sub_.subscribe(this, mask_topic_,
                        best_effort_qos.get_rmw_qos_profile());
    cloud_sub_.subscribe(this, pointcloud_topic_,
                         best_effort_qos.get_rmw_qos_profile());

    sync_ = std::make_shared<Synchronizer>(
        SyncPolicy(sync_queue_size), mask_sub_, cloud_sub_);
    sync_->setMaxIntervalDuration(rclcpp::Duration::from_seconds(sync_slop));
    sync_->registerCallback(std::bind(
        &SegmentedLidarClusterNode::syncedCallback, this, std::placeholders::_1,
        std::placeholders::_2));

    RCLCPP_INFO(this->get_logger(),
                "Segmented LiDAR clustering node started. mask=%s, cloud=%s, "
                "camera_info=%s",
                mask_topic_.c_str(), pointcloud_topic_.c_str(),
                camera_info_topic_.c_str());
  }

private:
  using MaskMsg = sensor_msgs::msg::CompressedImage;
  using PointCloud2Msg = sensor_msgs::msg::PointCloud2;
  using SyncPolicy =
      message_filters::sync_policies::ApproximateTime<MaskMsg, PointCloud2Msg>;
  using Synchronizer = message_filters::Synchronizer<SyncPolicy>;

  void cameraInfoCallback(const sensor_msgs::msg::CameraInfo::SharedPtr msg) {
    camera_info_ = msg;
  }

  void syncedCallback(const MaskMsg::ConstSharedPtr &mask_msg,
                      const PointCloud2Msg::ConstSharedPtr &cloud_msg) {
    if (!camera_info_) {
      RCLCPP_WARN_THROTTLE(this->get_logger(), *this->get_clock(), 2000,
                           "Waiting for CameraInfo.");
      return;
    }

    const std::string camera_frame =
        camera_frame_param_.empty() ? camera_info_->header.frame_id
                                    : camera_frame_param_;
    if (camera_frame.empty()) {
      RCLCPP_WARN(this->get_logger(),
                  "Camera frame is empty. Set camera_frame or publish "
                  "CameraInfo frame_id.");
      return;
    }

    cv::Mat mask = imageToBinaryMask(mask_msg);
    if (mask.empty()) {
      return;
    }

    geometry_msgs::msg::TransformStamped transform;
    try {
      transform = tf_buffer_.lookupTransform(
          camera_frame, cloud_msg->header.frame_id, cloud_msg->header.stamp,
          rclcpp::Duration::from_seconds(tf_timeout_sec_));
    } catch (const tf2::TransformException &ex) {
      RCLCPP_WARN_THROTTLE(
          this->get_logger(), *this->get_clock(), 2000,
          "Cannot transform cloud frame '%s' to camera frame '%s': %s",
          cloud_msg->header.frame_id.c_str(), camera_frame.c_str(), ex.what());
      return;
    }

    auto lidar_cloud = std::make_shared<pcl::PointCloud<pcl::PointXYZ>>();
    pcl::fromROSMsg(*cloud_msg, *lidar_cloud);

    if (lidar_cloud->empty()) {
      publishEmpty(cloud_msg->header);
      return;
    }

    const Eigen::Isometry3f lidar_to_camera =
        transformToEigen(transform.transform);
    auto selected_cloud = selectPointsOnMask(lidar_cloud, lidar_to_camera, mask);
    auto cluster_indices = clusterPoints(selected_cloud);
    auto centers = clusterCenters(selected_cloud, cluster_indices);

    publishFilteredCloud(cloud_msg->header, selected_cloud);
    publishCenters(cloud_msg->header, centers);
    publishMarkers(cloud_msg->header, selected_cloud, cluster_indices, centers);
  }

  cv::Mat imageToBinaryMask(const MaskMsg::ConstSharedPtr &mask_msg) {
    cv_bridge::CvImagePtr cv_ptr;
    try {
      cv_ptr = cv_bridge::toCvCopy(mask_msg);
    } catch (const cv_bridge::Exception &ex) {
      RCLCPP_WARN_THROTTLE(this->get_logger(), *this->get_clock(), 2000,
                           "Failed to convert mask image: %s", ex.what());
      return {};
    }

    cv::Mat gray;
    const cv::Mat &src = cv_ptr->image;
    if (src.channels() == 1) {
      gray = src;
    } else if (src.channels() == 3) {
      cv::cvtColor(src, gray, cv::COLOR_BGR2GRAY);
    } else if (src.channels() == 4) {
      cv::cvtColor(src, gray, cv::COLOR_BGRA2GRAY);
    } else {
      RCLCPP_WARN_THROTTLE(this->get_logger(), *this->get_clock(), 2000,
                           "Unsupported mask channel count: %d",
                           src.channels());
      return {};
    }

    cv::Mat gray_u8;
    if (gray.depth() == CV_8U) {
      gray_u8 = gray;
    } else {
      gray.convertTo(gray_u8, CV_8U);
    }

    cv::Mat binary;
    cv::threshold(gray_u8, binary, mask_threshold_, 255, cv::THRESH_BINARY);
    return binary;
  }

  pcl::PointCloud<pcl::PointXYZ>::Ptr selectPointsOnMask(
      const pcl::PointCloud<pcl::PointXYZ>::ConstPtr &lidar_cloud,
      const Eigen::Isometry3f &lidar_to_camera, const cv::Mat &mask) const {
    auto selected = std::make_shared<pcl::PointCloud<pcl::PointXYZ>>();
    selected->header = lidar_cloud->header;
    selected->is_dense = false;

    const double fx = camera_info_->k[0];
    const double fy = camera_info_->k[4];
    const double cx = camera_info_->k[2];
    const double cy = camera_info_->k[5];
    const double x_scale = camera_info_->width > 0
                               ? static_cast<double>(mask.cols) /
                                     static_cast<double>(camera_info_->width)
                               : 1.0;
    const double y_scale = camera_info_->height > 0
                               ? static_cast<double>(mask.rows) /
                                     static_cast<double>(camera_info_->height)
                               : 1.0;

    selected->points.reserve(lidar_cloud->points.size());
    for (const auto &point : lidar_cloud->points) {
      if (!pcl::isFinite(point)) {
        continue;
      }

      const Eigen::Vector3f lidar_point(point.x, point.y, point.z);
      const Eigen::Vector3f camera_point = lidar_to_camera * lidar_point;
      if (camera_point.z() <= 0.0F) {
        continue;
      }

      const int u = static_cast<int>(std::lround(
          (fx * camera_point.x() / camera_point.z() + cx) * x_scale));
      const int v = static_cast<int>(std::lround(
          (fy * camera_point.y() / camera_point.z() + cy) * y_scale));

      if (u < 0 || u >= mask.cols || v < 0 || v >= mask.rows) {
        continue;
      }
      if (mask.at<unsigned char>(v, u) == 0) {
        continue;
      }

      selected->points.push_back(point);
    }

    selected->width = static_cast<std::uint32_t>(selected->points.size());
    selected->height = 1;
    return selected;
  }

  std::vector<pcl::PointIndices>
  clusterPoints(const pcl::PointCloud<pcl::PointXYZ>::ConstPtr &cloud) const {
    std::vector<pcl::PointIndices> cluster_indices;
    if (static_cast<int>(cloud->points.size()) < min_cluster_size_) {
      return cluster_indices;
    }

    auto tree = std::make_shared<pcl::search::KdTree<pcl::PointXYZ>>();
    tree->setInputCloud(cloud);

    pcl::EuclideanClusterExtraction<pcl::PointXYZ> extractor;
    extractor.setClusterTolerance(cluster_tolerance_);
    extractor.setMinClusterSize(min_cluster_size_);
    extractor.setMaxClusterSize(max_cluster_size_);
    extractor.setSearchMethod(tree);
    extractor.setInputCloud(cloud);
    extractor.extract(cluster_indices);
    return cluster_indices;
  }

  std::vector<Eigen::Vector3f> clusterCenters(
      const pcl::PointCloud<pcl::PointXYZ>::ConstPtr &cloud,
      const std::vector<pcl::PointIndices> &cluster_indices) const {
    std::vector<Eigen::Vector3f> centers;
    centers.reserve(cluster_indices.size());

    for (const auto &cluster : cluster_indices) {
      Eigen::Vector4f centroid;
      pcl::compute3DCentroid(*cloud, cluster.indices, centroid);
      centers.emplace_back(centroid.x(), centroid.y(), centroid.z());
    }
    return centers;
  }

  void publishFilteredCloud(
      const std_msgs::msg::Header &header,
      const pcl::PointCloud<pcl::PointXYZ>::ConstPtr &cloud) const {
    PointCloud2Msg output;
    pcl::toROSMsg(*cloud, output);
    output.header = header;
    filtered_cloud_pub_->publish(output);
  }

  void publishCenters(const std_msgs::msg::Header &header,
                      const std::vector<Eigen::Vector3f> &centers) const {
    geometry_msgs::msg::PoseArray pose_array;
    pose_array.header = header;

    for (const auto &center : centers) {
      geometry_msgs::msg::Pose pose;
      pose.position.x = center.x();
      pose.position.y = center.y();
      pose.position.z = center.z();
      pose.orientation.w = 1.0;
      pose_array.poses.push_back(pose);
    }

    centers_pub_->publish(pose_array);
  }

  void publishMarkers(
      const std_msgs::msg::Header &header,
      const pcl::PointCloud<pcl::PointXYZ>::ConstPtr &cloud,
      const std::vector<pcl::PointIndices> &cluster_indices,
      const std::vector<Eigen::Vector3f> &centers) const {
    visualization_msgs::msg::MarkerArray marker_array;

    visualization_msgs::msg::Marker delete_marker;
    delete_marker.header = header;
    delete_marker.action = visualization_msgs::msg::Marker::DELETEALL;
    marker_array.markers.push_back(delete_marker);

    for (std::size_t i = 0; i < cluster_indices.size(); ++i) {
      Eigen::Vector4f min_point;
      Eigen::Vector4f max_point;
      pcl::getMinMax3D(*cloud, cluster_indices[i].indices, min_point,
                       max_point);

      const auto &center = centers[i];
      const double scale_x =
          std::max(static_cast<double>(max_point.x() - min_point.x()), 0.05);
      const double scale_y =
          std::max(static_cast<double>(max_point.y() - min_point.y()), 0.05);
      const double scale_z =
          std::max(static_cast<double>(max_point.z() - min_point.z()), 0.05);

      visualization_msgs::msg::Marker marker;
      marker.header = header;
      marker.ns = "segmented_lidar_clusters";
      marker.id = static_cast<int>(i);
      marker.type = visualization_msgs::msg::Marker::CUBE;
      marker.action = visualization_msgs::msg::Marker::ADD;
      marker.pose.position.x = center.x();
      marker.pose.position.y = center.y();
      marker.pose.position.z = center.z();
      marker.pose.orientation.w = 1.0;
      marker.scale.x = scale_x;
      marker.scale.y = scale_y;
      marker.scale.z = scale_z;
      marker.color.r = 0.1F;
      marker.color.g = 0.8F;
      marker.color.b = 1.0F;
      marker.color.a = 0.35F;
      marker_array.markers.push_back(marker);

      visualization_msgs::msg::Marker text_marker;
      text_marker.header = header;
      text_marker.ns = "segmented_lidar_cluster_labels";
      text_marker.id = static_cast<int>(i);
      text_marker.type = visualization_msgs::msg::Marker::TEXT_VIEW_FACING;
      text_marker.action = visualization_msgs::msg::Marker::ADD;
      text_marker.pose.position.x = center.x();
      text_marker.pose.position.y = center.y();
      text_marker.pose.position.z = center.z() + scale_z * 0.5 + 0.15;
      text_marker.pose.orientation.w = 1.0;
      text_marker.scale.z = 0.18;
      text_marker.color.r = 1.0F;
      text_marker.color.g = 1.0F;
      text_marker.color.b = 1.0F;
      text_marker.color.a = 1.0F;
      text_marker.text =
          std::to_string(i) + ": (" + formatFloat(center.x()) + ", " +
          formatFloat(center.y()) + ", " + formatFloat(center.z()) +
          ") n=" + std::to_string(cluster_indices[i].indices.size());
      marker_array.markers.push_back(text_marker);
    }

    markers_pub_->publish(marker_array);
  }

  void publishEmpty(const std_msgs::msg::Header &header) const {
    auto empty_cloud = std::make_shared<pcl::PointCloud<pcl::PointXYZ>>();
    publishFilteredCloud(header, empty_cloud);
    publishCenters(header, {});
    publishMarkers(header, empty_cloud, {}, {});
  }

  Eigen::Isometry3f
  transformToEigen(const geometry_msgs::msg::Transform &transform) const {
    Eigen::Isometry3f eigen_transform = Eigen::Isometry3f::Identity();
    eigen_transform.translation() =
        Eigen::Vector3f(transform.translation.x, transform.translation.y,
                        transform.translation.z);

    const Eigen::Quaternionf rotation(
        transform.rotation.w, transform.rotation.x, transform.rotation.y,
        transform.rotation.z);
    eigen_transform.linear() = rotation.normalized().toRotationMatrix();
    return eigen_transform;
  }

  std::string formatFloat(float value) const {
    char buffer[32];
    std::snprintf(buffer, sizeof(buffer), "%.2f", value);
    return std::string(buffer);
  }

  std::string mask_topic_;
  std::string pointcloud_topic_;
  std::string camera_info_topic_;
  std::string filtered_cloud_topic_;
  std::string centers_topic_;
  std::string markers_topic_;
  std::string camera_frame_param_;
  int mask_threshold_{0};
  double cluster_tolerance_{0.30};
  int min_cluster_size_{8};
  int max_cluster_size_{30000};
  double tf_timeout_sec_{0.05};

  sensor_msgs::msg::CameraInfo::SharedPtr camera_info_;
  rclcpp::Subscription<sensor_msgs::msg::CameraInfo>::SharedPtr
      camera_info_sub_;
  rclcpp::Publisher<PointCloud2Msg>::SharedPtr filtered_cloud_pub_;
  rclcpp::Publisher<geometry_msgs::msg::PoseArray>::SharedPtr centers_pub_;
  rclcpp::Publisher<visualization_msgs::msg::MarkerArray>::SharedPtr
      markers_pub_;

  message_filters::Subscriber<MaskMsg> mask_sub_;
  message_filters::Subscriber<PointCloud2Msg> cloud_sub_;
  std::shared_ptr<Synchronizer> sync_;

  tf2_ros::Buffer tf_buffer_;
  tf2_ros::TransformListener tf_listener_;
};

int main(int argc, char **argv) {
  rclcpp::init(argc, argv);
  rclcpp::spin(std::make_shared<SegmentedLidarClusterNode>());
  rclcpp::shutdown();
  return 0;
}
